File: 2371-minimize-maximum-value-in-a-grid/2371-minimize-maximum-value-in-a-grid.java
class Solution {
    public int[][] minScore(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        List<int[]> nums = new ArrayList<>();
        for(int i = 0; i < m; ++i)
            for(int j = 0; j < n; ++j)
                nums.add(new int[]{grid[i][j], i, j});
        Collections.sort(nums, (a, b) -> a[0] - b[0]);
        
        int[] rm = new int[m];
        int[] cm = new int[n];
        int[][] ans = new int[m][n];
        
        for(int[] num : nums) {
            int i = num[1]; int j = num[2];
            ans[i][j] = Math.max(rm[i], cm[j]) + 1;
            rm[i] = cm[j] = ans[i][j];
        }
        
        return ans;
    }
}

File: 2371-minimize-maximum-value-in-a-grid/2371-minimize-maximum-value-in-a-grid.py
class Solution(object):
    def minScore(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: List[List[int]]
        """
        m, n = len(grid), len(grid[0])
        nums = [(v, i, j) for (i, row) in enumerate(grid) for (j, v) in enumerate(row)]
        nums.sort()
        
        rm = [0] * m
        cm = [0] * n
        ans = [[0] * n for _ in range(m)]
        
        for num in nums:
            i, j = num[1], num[2]
            ans[i][j] = max(rm[i], cm[j]) + 1
            rm[i] = cm[j] = ans[i][j]
            
        return ans